deeplearning4j-core/src/test/java/org/deeplearning4j/gradientcheck/CNNGradientCheckTest.java

```diff
@@ -645,35 +646,70 @@ public void testCnnZeroPaddingLayer() {
                     }
 
                     boolean gradOK = GradientCheckUtil.checkGradients(net, DEFAULT_EPS, DEFAULT_MAX_REL_ERROR,
-                                    DEFAULT_MIN_ABS_ERROR, PRINT_RESULTS, RETURN_ON_FIRST_FAILURE, input, labels);
+                            DEFAULT_MIN_ABS_ERROR, PRINT_RESULTS, RETURN_ON_FIRST_FAILURE, input, labels);
 
                     assertTrue(msg, gradOK);
                 }
             }
         }
     }
 
+    @Test(expected = IllegalArgumentException.class)
+    public void testDeconvolution2DUnsupportedSameModeLayer() {
+        /*
+         * When setting border mode same on layer directly, we can catch
+         * it in the builder
+         */
+        Deconvolution2D.Builder deconv = new Deconvolution2D.Builder();
+        deconv.convolutionMode(ConvolutionMode.Same);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testDeconvolution2DUnsupportedSameModeNetwork() {
+        /*
+         * When convolution mode Same is set for the network and a deconvolution layer is added
+         * then only layer activation will fail. Suboptimal, but I don't think we want special
+         * logic for NNC in this case.
+         */
+        NeuralNetConfiguration.ListBuilder b = new NeuralNetConfiguration.Builder().seed(12345)
+                .updater(new NoOp())
+                .activation(Activation.SIGMOID)
+                .convolutionMode(ConvolutionMode.Same)
+                .list()
+                .layer(new Deconvolution2D.Builder().name("deconvolution")
+                        .nIn(3).nOut(2).build());
+
+        MultiLayerConfiguration conf = b.layer(new OutputLayer.Builder(LossFunctions.LossFunction.MCXENT)
+                .activation(Activation.SOFTMAX).nOut(2).build())
+                .setInputType(InputType.convolutionalFlat(7, 7, 3)).build();
+
+        MultiLayerNetwork net = new MultiLayerNetwork(conf);
+        net.init();
+        net.getLayer(0).activate(Nd4j.rand(10, 7 * 7 * 3));
+    }
+
     @Test
-    public void testSeparableConv2D() {
+    public void testDeconvolution2D() {
         int nOut = 2;
 
         int minibatchSize = 3;
```

Comment:
Minibatch size 1 is an important edge case (I've seen bugs appear only on that before).
Maybe check both 1 and 3